vm/src/stdlib/bz2.rs

```diff
@@ -0,0 +1,213 @@
+    Ok(BZ2Decompressor {
+        state: PyMutex::new(DecompressorState {
+            decoder: Decompress::new(false),
+            // TODO: too small?
+            buf: Vec::with_capacity(32 * 1024),
+            unused_data: Vec::with_capacity(32 * 1024),
+            eof: false,
+            needs_input: true,
+        }),
+    })
+}
+
+struct CompressorState {
+    flushed: bool,
+    encoder: Option<BzEncoder<Vec<u8>>>,
+}
+
+#[pyclass(module = "_bz2", name = "BZ2Compressor")]
+struct BZ2Compressor {
+    state: PyMutex<CompressorState>,
+}
+
+impl fmt::Debug for BZ2Compressor {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "_bz2.BZ2Compressor")
+    }
+}
+
+impl PyValue for BZ2Compressor {
+    fn class(_vm: &VirtualMachine) -> &PyTypeRef {
+        Self::static_type()
+    }
+}
+
+// TODO: return partial results from compress() instead of returning everything in flush()
+#[pyimpl]
+impl BZ2Compressor {
+    #[pymethod]
+    fn compress(&self, data: PyBytesLike, vm: &VirtualMachine) -> PyResult {
+        let mut state = self.state.lock();
+        if state.flushed {
+            return Err(vm.new_value_error("Compressor has been flushed".to_owned()));
+        }
+
+        let CompressorState { flushed, encoder } = &mut *state;
+
+        // TODO: handle Err
+        data.with_ref(|input_bytes| encoder.as_mut().unwrap().write_all(input_bytes).unwrap());
+        Ok(vm.ctx.new_bytes(Vec::new()))
+    }
+
+    #[pymethod]
+    fn flush(&self, vm: &VirtualMachine) -> PyResult {
+        let mut state = self.state.lock();
+        if state.flushed {
+            return Err(vm.new_value_error("Repeated call to flush()".to_owned()));
+        }
+
+        let CompressorState { flushed, encoder } = &mut *state;
+
+        // TODO: handle Err
+        let out = encoder.take().unwrap().finish().unwrap();
+        state.flushed = true;
+        Ok(vm.ctx.new_bytes(out.to_vec()))
+    }
+}
+
+fn _bz2_BZ2Compressor(
+    compresslevel: OptionalArg<i32>, // TODO: PyIntRef
+    vm: &VirtualMachine,
+) -> PyResult<BZ2Compressor> {
+    // TODO: this library should use constants instead
+    let compresslevel = compresslevel.unwrap_or(9);
+    // compresslevel.unwrap_or(bzip2::Compression::best().level().try_into().unwrap());
+    let level = match compresslevel {
+        valid_level @ 1..=9 => bzip2::Compression::new(valid_level as u32),
+        _ => return Err(vm.new_value_error("compresslevel must be between 1 and 9".to_owned())),
+    };
+
+    Ok(BZ2Compressor {
+        state: PyMutex::new(CompressorState {
+            flushed: false,
+            encoder: Some(BzEncoder::new(Vec::new(), level)),
+        }),
+    })
+}
+
+pub fn make_module(vm: &VirtualMachine) -> PyObjectRef {
+    let ctx = &vm.ctx;
+    BZ2Decompressor::make_class(ctx);
+    BZ2Compressor::make_class(ctx);
+
+    py_module!(vm, "_bz2", {
+        "BZ2Decompressor" => named_function!(ctx, _bz2, BZ2Decompressor),
```

Comment:
I know this is a WIP, but just a note - it looks like CPython exports `_bz2.BZ2[De]compressor` as the classes, rather as separate functions. You can move the functionality from `_bz2_BZ2Decompressor` to an associated function `BZ2Decompressor::tp_new`, with a `#[pyslot]` attribute on it, and same with `BZ2Compressor`. Then you can export directly as `"BZ2Decompressor" => BZ2Decompressor::make_class(ctx)`